### PR TITLE
ADBDEV-6062: Fix tablespace mapping during restore

### PR DESCRIPTION
**Pull request by bandetto**
Fix tablespace mapping during restore

Previously, if --tablespace-map/--tablespace-map-all flag was used to restore a
GPDB segment's data, the symlink inside pg_tblspc for remapped tablespaces was
incorrectly pointing to \<tablespace_dir\>/ instead of
\<tablespace_dir\>/\<dbid\>.

The dbid part is actually never stored anywhere in the manifest or stanza.
Instead, tablespace path is put as an absolute path with the dbid as the base
folder. So the way used to obtain dbid is to get the basename of the
GPDB-specific tablespace path.

This patch adds GPDB-only logic for tablespace remapping to include dbid part
back into the path of restored tablespaces.

### CODE REVIEW

File: src/command/restore/restore.c [hunk 1]
```diff
@@ -528,10 +531,16 @@ restoreManifestMap(Manifest *const manifest)
                     // Remap tablespace if a mapping was found
                     if (tablespacePath != NULL)
                     {
-                        LOG_INFO_FMT("map tablespace '%s' to '%s'", strZ(target->name), strZ(tablespacePath));
+                        // Append dbid to the path for Greenplum
+                        const String *const fixedTablespacePath =
+                            cfgOptionStrId(cfgOptFork) == CFGOPTVAL_FORK_GPDB ?
+                                strNewFmt("%s/%u", strZ(tablespacePath), manifestData(manifest)->pgId) : tablespacePath;
+
+                        LOG_INFO_FMT("map tablespace '%s' to '%s'", strZ(target->name), strZ(fixedTablespacePath));
```
Comment (KnightMurloc) [inline src/command/restore/restore.c]: Maybe use if instead of the trinary operator? I think it will be more readable.
It may also be worth assigning a new value to the original variable, this will make the diff smaller.

File: src/command/restore/restore.c [hunk 2]
```diff
@@ -528,10 +531,16 @@ restoreManifestMap(Manifest *const manifest)
                     // Remap tablespace if a mapping was found
                     if (tablespacePath != NULL)
                     {
-                        LOG_INFO_FMT("map tablespace '%s' to '%s'", strZ(target->name), strZ(tablespacePath));
+                        // Append dbid to the path for Greenplum
+                        const String *const fixedTablespacePath =
+                            cfgOptionStrId(cfgOptFork) == CFGOPTVAL_FORK_GPDB ?
+                                strNewFmt("%s/%u", strZ(tablespacePath), manifestData(manifest)->pgId) : tablespacePath;
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: Do we really need new variable?
```suggestion
                        if (cfgOptionStrId(cfgOptFork) == CFGOPTVAL_FORK_GPDB)
                                tablespacePath = strNewFmt("%s/%u", strZ(tablespacePath), manifestData(manifest)->pgId);
```

File: src/command/restore/restore.c [hunk 3]
```diff
@@ -485,8 +485,11 @@ restoreManifestMap(Manifest *const manifest)
                 // Is this a tablespace?
                 if (target->tablespaceId != 0)
                 {
+                    // Tablespace path without dbid
                     const String *tablespacePath = NULL;
 
+                    LOG_INFO_FMT("using tablespace '%s' in '%s'", strZ(target->name), strZ(target->path));
```
Comment (KnightMurloc) [inline src/command/restore/restore.c]: Do we need this log?

File: src/command/restore/restore.c (hunk 1 above)
Comment (RekGRpth) [reply src/command/restore/restore.c]: https://github.com/arenadata/pgbackrest/pull/38#discussion_r1709195680 ;)

Comment (bandetto) [reply src/command/restore/restore.c]: Sure

File: src/command/restore/restore.c (hunk 3 above)
Comment (bandetto) [reply src/command/restore/restore.c]: Currently, the log does not show the previous tablespace path, only the new one. I found this helpful.

File: src/command/restore/restore.c [hunk 4]
```diff
@@ -528,10 +531,15 @@ restoreManifestMap(Manifest *const manifest)
...
                     if (tablespacePath != NULL)
                     {
+                        // Append dbid to the path for Greenplum
+                        if (cfgOptionStrId(cfgOptFork) == CFGOPTVAL_FORK_GPDB)
+                            tablespacePath = strNewFmt("%s/%u", strZ(tablespacePath), manifestData(manifest)->pgId);
+
                         LOG_INFO_FMT("map tablespace '%s' to '%s'", strZ(target->name), strZ(tablespacePath));
 
                         manifestTargetUpdate(manifest, target->name, tablespacePath, NULL);
-                        manifestLinkUpdate(manifest, strNewFmt(MANIFEST_TARGET_PGDATA "/%s", strZ(target->name)), tablespacePath);
```
Comment (RekGRpth) [inline src/command/restore/restore.c]: Why is this line changed?

File: src/command/restore/restore.c (hunk 3 above)
Comment (RekGRpth) [reply src/command/restore/restore.c]: may be change log with map adding old path to it?

Comment (KnightMurloc) [reply src/command/restore/restore.c]: Maybe output a new log only for gpdb with information about how it was changed?

Comment (RekGRpth) [reply src/command/restore/restore.c]: Does the path change in postgres?

Comment (bandetto) [reply src/command/restore/restore.c]: The path is changed both in Postgres and GPDB, but GPDB additionally requires the dbid part. I think it's ok to include previous path in the logs for both variants.

**Comment by KnightMurloc** [comment]
I am faced with the following problem. It looks like in backup.info all segments have a dbid of 1, which leads to an attempt to restore the tablespace of all segments to the same directory. Is this the expected behavior? I did a backup according to the instructions from the repository by adding `--tablespace-map-all /tmp/thisisfine` to the recovery command.

File: src/command/restore/restore.c [hunk 5]
```diff
@@ -528,7 +530,28 @@ restoreManifestMap(Manifest *const manifest)
...
-                        LOG_INFO_FMT("map tablespace '%s' to '%s'", strZ(target->name), strZ(tablespacePath));
+                        // We'll need to append dbid to the path for Greenplum. Currently, nothing really stores a correct dbid of
+                        // a segment that was used, so just retrieve the last folder as a segment number.
+                        if (cfgOptionStrId(cfgOptFork) == CFGOPTVAL_FORK_GPDB)
+                        {
+                            const char *const dbid = strBaseZ(target->path);
+
+                            TRY_BEGIN()
+                            {
+                                // Try to convert dbid to a number and explode if it isn't one.
```
Comment (KnightMurloc) [inline src/command/restore/restore.c]: Maybe?
```suggestion
                                // Try to convert dbid to a number and throw an error if it isn't one.
```

Comment (bandetto) [reply src/command/restore/restore.c]: But it's more fun and the catch block consists only of one THROW call so it's easy to see :c

Comment (KnightMurloc) [reply src/command/restore/restore.c]: OK

**Review by KnightMurloc — DISMISSED** [review]

File: src/command/restore/restore.c [hunk 6]
```diff
@@ -485,6 +486,7 @@ restoreManifestMap(Manifest *const manifest)
                 // Is this a tablespace?
                 if (target->tablespaceId != 0)
                 {
+                    // Tablespace path without dbid
```
Comment (andr-sokolov) [inline src/command/restore/restore.c]: I suggest to delete this comment. The patch should be smaller if possible, because we are not upstream

File: src/command/restore/restore.c [hunk 7]
```diff
@@ -528,7 +530,28 @@ restoreManifestMap(Manifest *const manifest)
...
+                        // We'll need to append dbid to the path for Greenplum. Currently, nothing really stores a correct dbid of
+                        // a segment that was used, so just retrieve the last folder as a segment number.
+                        if (cfgOptionStrId(cfgOptFork) == CFGOPTVAL_FORK_GPDB)
+                        {
+                            const char *const dbid = strBaseZ(target->path);
+
+                            TRY_BEGIN()
+                            {
+                                // Try to convert dbid to a number and explode if it isn't one.
+                                tablespacePath = strNewFmt("%s/%d", strZ(tablespacePath), cvtZToInt(dbid));
```
Comment (andr-sokolov) [inline src/command/restore/restore.c]: Why dbid is converted from string to number and then immediately from number to string? Can we append dbid to tablespacePath without two converstions?

File: src/command/restore/restore.c [hunk 8]
```diff
@@ -528,7 +530,28 @@ restoreManifestMap(Manifest *const manifest)
...
+                            }
+                            CATCH_ANY()
+                            {
+                                THROW_FMT(
+                                    TablespaceMapError,
+                                    "greenplum-specific tablespace path '%s' contains invalid dbid directory '%s'",
```
Comment (andr-sokolov) [inline src/command/restore/restore.c]: What is the case, when we reach this line? Is the case expected? Can we remove the try-catch?

File: test/src/module/command/restoreTest.c [hunk 9]
```diff
@@ -814,6 +814,8 @@ testRun(void)
         TEST_ERROR(
             restoreManifestMap(manifest), TablespaceMapError, "tablespace remapped by name 'ts2' and id 2 with different paths");
 
+        // Postgres tests first
```
Comment (andr-sokolov) [inline test/src/module/command/restoreTest.c]: I suggest to delete this comment to make the patch smaller

File: test/src/module/command/restoreTest.c [hunk 10]
```diff
@@ -832,8 +834,8 @@ testRun(void)
             manifestLinkFind(manifest, STRDEF("pg_data/pg_tblspc/2"))->destination, "/2-2", "check tablespace 1 link");
 
         TEST_RESULT_LOG(
-            "P00   INFO: map tablespace 'pg_tblspc/1' to '/1-2'\n"
-            "P00   INFO: map tablespace 'pg_tblspc/2' to '/2-2'");
+            "P00   INFO: map tablespace 'pg_tblspc/1' from '/1' to '/1-2'\n"
+            "P00   INFO: map tablespace 'pg_tblspc/2' from '/2' to '/2-2'");
```
Comment (andr-sokolov) [inline test/src/module/command/restoreTest.c]: Can we avoid this change here and below?

File: src/command/restore/restore.c (hunk 8 above)
Comment (bandetto) [reply src/command/restore/restore.c]: This case is covered in tests. In case of Greenplum, tablespace link should point to the dbid directory, which should always be a number

File: src/command/restore/restore.c (hunk 7 above)
Comment (bandetto) [reply src/command/restore/restore.c]: The comment above the change clearly explains the purpose. dbid folder name should be a number

File: test/src/module/command/restoreTest.c (hunk 10 above)
Comment (bandetto) [reply test/src/module/command/restoreTest.c]: I think this change is helpful and it's unlikely that this part of the code will be affected by merge

File: src/command/restore/restore.c (hunk 6 above)
Comment (bandetto) [reply src/command/restore/restore.c]: I think it's confusing that the tablespace path does not have a dbid in this variable without the comment

File: src/command/restore/restore.c [hunk 11]
```diff
@@ -528,6 +528,11 @@ restoreManifestMap(Manifest *const manifest)
                     // Remap tablespace if a mapping was found
                     if (tablespacePath != NULL)
                     {
+                        // We'll need to append dbid to the path for Greenplum. Currently, nothing really stores a correct dbid of
+                        // a segment that was used, so just retrieve the last folder as a segment number.
+                        if (cfgOptionStrId(cfgOptFork) == CFGOPTVAL_FORK_GPDB)
+                            tablespacePath = strNewFmt("%s/%s", strZ(tablespacePath), strBaseZ(target->path));
```
Comment (andr-sokolov) [inline src/command/restore/restore.c:534]: What about using strCatChr+strCatZ or strCatFmt to don't allocate memory for a new String structure and don't copy the tablespacePath content?

Comment (bandetto) [reply src/command/restore/restore.c:534]: Then we would need to remove `const` from `tablespacePath` declaration, adding an extra hunk to changes.
```
error: passing argument 1 of 'strCatFmt' discards 'const' qualifier from pointer target type
```

**Review by andr-sokolov — DISMISSED** [review]

File: test/src/module/command/restoreTest.c [hunk 12]
```diff
@@ -876,6 +876,91 @@ testRun(void)
...
+        hrnCfgArgRaw(argList, cfgOptRepoPath, repoPath);
+        hrnCfgArgRaw(argList, cfgOptPgPath, pgPath);
+        hrnCfgArgRawZ(argList, cfgOptTablespaceMap, "3=/gpdb_1-2");
+        hrnCfgArgRawZ(argList, cfgOptTablespaceMap, "gpdb_ts2=/gpdb_2-2");
+
+        hrnCfgArgRawZ(argList, cfgOptFork, "GPDB");
+        HRN_CFG_LOAD(cfgCmdRestore, argList);
+
+        TEST_RESULT_VOID(restoreManifestMap(manifest), "remap tablespaces");
+        TEST_RESULT_STR_Z(manifestTargetFind(manifest, STRDEF("pg_tblspc/gpdb_1"))->path, "/gpdb_1-2/0", "check tablespace 1 target");
```
Comment (KnightMurloc) [inline test/src/module/command/restoreTest.c]: The length of the line must not exceed 132 characters. Many lines like this one in the test began to exceed this limit. The strange thing is that the code formatter did not detect this.

**Review by KnightMurloc — APPROVED** [review]

**Review by andr-sokolov — APPROVED** [review]
